### PR TITLE
Fix ann_test.py numerical bug in target reshape

### PR DESCRIPTION
**Pull request by gulsumgudukbay**
The reshape call has a numerical bug where the target reshape size is always higher than array size if num_devices is not a nonnegative power of two. To fix this, this commit adds padding.

### CODE REVIEW

File: tests/ann_test.py [hunk 1]
```diff
@@ -125,7 +125,23 @@ def test_pmap(self, qy_shape, db_shape, dtype, k, recall):
     db_size = db.shape[0]
     gt_scores = lax.dot_general(qy, db, (([1], [1]), ([], [])))
     _, gt_args = lax.top_k(-gt_scores, k)  # negate the score to get min-k
-    db_per_device = db_size//num_devices
```
Comment (jakevdp) [inline tests/ann_test.py:128]: Could we just `assert num_devices % 2 == 0` instead?

Comment (gulsumgudukbay) [reply tests/ann_test.py:128]: In the cases that if a machine contains any number of devices that is not a power of two, the assertion will fail and test will not proceed. If we want to test on those kinds of machines, the functionality won't be tested properly.

Comment (jakevdp) [reply tests/ann_test.py:128]: Do we test on such machines?

Comment (gulsumgudukbay) [reply tests/ann_test.py:128]: > Do we test on such machines?

We test on such machines.

Comment (jakevdp) [reply tests/ann_test.py:128]: Why has the failure not been caught before now?

I'm asking this because I think the appropriate fix may be to skip if the number of devices is not a multiple of 2. All this additional logic added in this PR does not seem in the spirit of the original test.

Comment (jakevdp) [reply tests/ann_test.py:128]: (also `pmap` is deprecated so it's not all that important to improve its test coverage for rare device geometries).

Comment (gulsumgudukbay) [reply tests/ann_test.py:128]: I think probably previously it was not tested on a machine where number of GPUs is not a power of 2 (we caught it when we tested it on a 7-GPU machine). We found another test that has this issue: tests/ffi_test.py test_shard_map.

Comment (jakevdp) [reply tests/ann_test.py:128]: OK, so it's not covered by CI which is why we haven't seen this error before.

What would you think about just adding `self.skipTest()` when `num_devices` is not a power of 2?

Comment (gulsumgudukbay) [reply tests/ann_test.py:128]: > OK, so it's not covered by CI which is why we haven't seen this error before.
> 
> What would you think about just adding `self.skipTest()` when `num_devices` is not a power of 2?

sure I can do that

Comment (gulsumgudukbay) [reply tests/ann_test.py:128]: @jakevdp  pushed another commit for the skip.

File: tests/ann_test.py [hunk 2]
```diff
@@ -119,6 +119,8 @@ def test_autodiff(self, shape, dtype, k, is_max_k):
   )
   def test_pmap(self, qy_shape, db_shape, dtype, k, recall):
     num_devices = jax.device_count()
+    if num_devices % 2 != 0:
+      self.skipTest("Works only when number of devices is a power of 2.")
```
Comment (jakevdp) [inline tests/ann_test.py]: nit - I know I said "power of 2" but it should be "multiple of 2" I think.

Comment (gulsumgudukbay) [reply tests/ann_test.py]: fixed. Let me know if anything else need to be changed.

**Review by jakevdp — COMMENTED** [review]
Looks good, thanks! Last thing: can you please squash the changes into a single commit? See https://docs.jax.dev/en/latest/contributing.html#single-change-commits-and-pull-requests.

Thanks!

**Comment by gulsumgudukbay** [comment]
> Looks good, thanks! Last thing: can you please squash the changes into a single commit? See https://docs.jax.dev/en/latest/contributing.html#single-change-commits-and-pull-requests.
> 
> Thanks!

Hi @jakevdp  I squashed the changes. Please let me know if I need to do any other change.

**Review by jakevdp — APPROVED** [review]
Looks good!
